util.c

```diff
@@ -1932,6 +1932,55 @@ int mdmon_running(const char *devnm)
 	return 0;
 }
 
+/* wait_for_mdmon_control_socket() - Waits for mdmon control socket
+ * to be created within specified time.
+ * @devnm: Device for which mdmon control socket should start.
+ *
+ * In foreground mode, when mdadm is trying to connect to control
+ * socket it is possible that the mdmon has not created it yet.
+ * Give some time to mdmon to create socket. Timeout set to 10 sec.
+ */
+int wait_for_mdmon_control_socket(char *devname)
+{
+	char path[100];
+	int sfd;
+	struct sockaddr_un addr;
+	int pos;
+	char *c;
+	int rv, retry_count = 0;
+	pos = sprintf(path, "%s/", MDMON_DIR);
+	if (is_subarray(devname)) {
+		devname++;
+		c = strchr(devname, '/');
+		if (!c)
+			return -1;
+		snprintf(&path[pos], c - devname + 1, "%s", devname);
+		pos += c - devname;
+	} else
+		pos += sprintf(&path[pos], "%s", devname);
+	sprintf(&path[pos], ".sock");
+
+	sfd = socket(PF_LOCAL, SOCK_STREAM, 0);
+	if (sfd < 0)
+		return -1;
+
+	addr.sun_family = PF_LOCAL;
+	strcpy(addr.sun_path, path);
+
+	for (retry_count = 0; retry_count < 20; retry_count++) {
```

Comment:
I had internal discussion with my team, and we considered a scenario where system is running heavy workload or is under heavy stress, in that case it is possible that the mdmon can take some time to create a socket. That's why I changed the time back to 10 seconds.

Also as far as I remember the udev kill a process/thread if it didn't get a response for more than 180 seconds. Please correct me if you are referring to something else.

Reply:
Here a man:
https://man7.org/linux/man-pages/man7/udev.7.html
```
 IMPORT{type}
           Import a set of variables as device properties, depending on
           type:

           "program"
               Execute an external program specified as the assigned
               value and, if it returns successfully, import its output,
               which must be in environment key format. Path
               specification, command/argument separation, and quoting
               work like in RUN.

 {...}

           This can only be used for very short-running foreground
           tasks. For details see RUN.

```

It is not a case for create for sure, create is done from cmdline. But in a case that at some point I would like to reuse this logic in Incremental then I may get a problem, because we have:

```
ACTION!="remove", IMPORT{program}="BINDIR/mdadm --incremental --export $devnode --offroot $env{DEVLINKS}"
```
I can see that in `assemble_container_content` we are pinging mdmon in case of reshape so the question is if we should change that to the method you proposed. And here is why I would like to avoid long running task.

Udev is not strict what are "short-running foreground tasks". I expect seconds not minutes to be allowed. But this is still theoretical issue as we expect mdmon to be started fast.

So I would stay that if there is no real issue we are familiar with that requires to wait longer for mdmon, we should stay with 2 seconds unless you can really prove that more than 10 seconds is safe.

Reply:
> What if we pass timeout value as argument to the function? This way we can have different timeout value for different sceanrios.

Has some sense but we cannot expect from future developers to know what is reasonable. Also, please ensure that the research you are doing will not be forgotten. I want to respect the time you have spent on analyzing it.

So, please prepare some predefined values that developers may reuse without additional analysis. For example, by using `bool udev_frendly_wait` and defining long and shot time that is reasonable in your opinion. I will trust your judgement here.

 Sometimes `bool force` is used for similar purposes in mdadm. I have no strong preference for the name but please comment it well.

Reply:
Lets do one thing, I will switch back to 2 second delay and I can add an error print if connect fails even after 2 seconds. 
And in future if we see connect failing in wait_for_mdmon_control_socket() for cases where we dont want it to fail then we can work on increasing the timeout limit along with all the fail safe for udev.

Reply:
I have changed the sleep timeout to 2 second and have also added a debug print in case timout occur in wait_for_mdmon_control_socket().
Please review it.

Reply:
Thanks, will look into it tomorrow!

Reply:
A gentle reminder.